### PR TITLE
Add remaining data type support in dictionary

### PR DESCRIPTION
**Pull request by tboonIV**
add data type support in dictionary - blob, bufferId, fieldKey, objectId and phoneno
update XML file loader to support iso-8859-1 encoding

### CODE REVIEW

**Comment by tboonIV** [comment]
working on the failure tests

File: cmdc/value.go [hunk 1]
```diff
@@ -354,3 +356,45 @@ func decodeTimeValue(b []byte) (*time.Time, error) {
...
+			return "", fmt.Errorf("bad format: '%s' is not a valid phone number", value)
+		}
+	}
+	return value, nil
+}
+
+func encodeObjectIDValue(v string) ([]byte, error) {
+	return []byte(v), nil
+}
+
```
Comment (lwlee2608) [inline cmdc/value.go:383]: This objectId format (eg. 1-2-3-100) is not cmdc specific and, therefore should reside in `mdd/field` package. Same goes with `phoneNo` as well

File: dictionary/dictionary.go [hunk 2]
```diff
@@ -160,6 +196,7 @@ func (d *Dictionary) Lookup(key, schemaVersion, extVersion int) (*ContainerDefin
 			d.Add(result)
 			return result, true
 		}
+		log.Errorf("Fail to lookup %v, %v", ckey, err)
```
Comment (lwlee2608) [inline dictionary/dictionary.go:199]: Remove logging here. We should rely more on better error handling instead

File: dictionary/parser.go [hunk 3]
```diff
@@ -50,15 +52,12 @@ type Field struct {
...
-	}
+	decoder := xml.NewDecoder(reader)
+	decoder.CharsetReader = charset.NewReaderLabel
 
 	var config Configuration
-	err = xml.Unmarshal(data, &config)
-	if err != nil {
-		return nil, err
+	if err := decoder.Decode(&config); err != nil {
+		panic(err)
```
Comment (lwlee2608) [inline dictionary/parser.go:60]: return error instead of panic

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+bin/
```
Comment (lwlee2608) [inline .gitignore:1]: why would you have these two directories?

File: dictionary/dictionary.go [hunk 5]
```diff
@@ -194,12 +193,12 @@ func (d *Dictionary) Lookup(key, schemaVersion, extVersion int) (*ContainerDefin
 		result, err := d.search(key, schemaVersion, extVersion)
 		if err == nil {
 			d.Add(result)
-			return result, true
+			return result, true, nil
 		}
-		log.Errorf("Fail to lookup %v, %v", ckey, err)
+		return nil, found, err
```
Comment (lwlee2608) [inline dictionary/dictionary.go:198]: The new `error` return seems unnecessary. Didn't the bool `found` already indicate the same thing?

File: mdd/mdd.go [hunk 6]
```diff
@@ -7,6 +7,7 @@ import (
 
 	"github.com/matrixxsoftware/go-mdd/dictionary"
 	"github.com/matrixxsoftware/go-mdd/mdd/field"
+	"google.golang.org/appengine/log"
```
Comment (lwlee2608) [inline mdd/mdd.go:10]: What is this?

Comment (tboonIV) [reply mdd/mdd.go:10]: used wrong log. fixing it

File: dictionary/dictionary.go (hunk 5 above)
Comment (tboonIV) [reply dictionary/dictionary.go:198]: the error might cause by invalid data type, if we didn't return error for logging, would it actual error get hidden?

File: .gitignore (hunk 4 above)
Comment (tboonIV) [reply .gitignore:1]: yea, I shouldn't have this 2 directories. let me remove this file

File: dictionary/dictionary.go (hunk 5 above)
Comment (tboonIV) [reply dictionary/dictionary.go:198]: I want to print error message look like this that have a bit more detail. 
```
ERRO[2024-07-19T06:20:27Z] [10.244.1.7:14061->10.244.1.1:17320] Error lookup definition key 447 schemaVersion 5263 extVersion 1: Error field SubscriberId: Unknown datatype: object id                                                                        
INFO[2024-07-19T06:20:57Z] [10.244.1.7:14061->10.244.1.1:17320] Connection closed
```
maybe I can remove the bool and replace with error

Comment (lwlee2608) [reply dictionary/dictionary.go]: ok make sense to replace `found` bool with `err`

File: mdd/mdd.go [hunk 7]
```diff
@@ -332,6 +332,32 @@ func NewNullField(fieldType field.Type) *Field {
 	}
 }
 
+func NewObjectIDField(value string) *Field {
+	objectIDValue, err := field.NewObjectID([]byte(value))
+	if err != nil {
+		return NewNullField(field.ObjectID)
```
Comment (lwlee2608) [inline mdd/mdd.go:338]: This is bad, should not default to null field if the value format is wrong.

File: mdd/mdd.go [hunk 8]
```diff
@@ -333,6 +332,32 @@ func NewNullField(fieldType field.Type) *Field {
...
+	}
+	return &Field{
+		Type:        field.ObjectID,
+		Value:       objectIDValue,
+		IsMulti:     false,
+		IsContainer: false,
+	}
+}
+
+func NewPhoneNoField(value string) *Field {
```
Comment (lwlee2608) [inline mdd/mdd.go:348]: remove this two new functions `NewPhoneNoField` and `NewObjectIdField`

File: mdd/mdd.go [hunk 9]
```diff
@@ -52,14 +53,15 @@ func (c *Containers) GetContainer(key int) *Container {
...
-		definition, ok := definitions.Lookup(
+		definition, err := definitions.Lookup(
 			container.Header.Key,
 			container.Header.SchemaVersion,
 			container.Header.ExtVersion,
 		)
-		if ok {
-			container.LoadDefinition(definition)
+		if err != nil {
+			log.Errorf("Error lookup: %v", err)
```
Comment (lwlee2608) [inline mdd/mdd.go:62]: return error instead of logging error

File: mdd/field/objectid.go [hunk 10]
```diff
@@ -5,9 +5,10 @@ import (
 	"unicode"
 )
```
[the remark this answers is not in the document]
Comment (tboonIV) [reply mdd/field/objectid.go:7]: crash with type `ObjectID` in type.go

[the remark this answers is not in the document]
Comment (lwlee2608) [reply mdd/field/objectid.go:7]: oic..

**Review by lwlee2608 — APPROVED** [review]

**Pull request merged**